src/errors/types.rs

```diff
@@ -58,263 +58,355 @@ pub fn list_all_errors(py: Python) -> PyResult<&PyList> {
     Ok(PyList::new(py, errors))
 }
 
-/// Definite each validation error.
-/// NOTE: if an error has parameters:
-/// * the variables in the message need to match the enum struct
-/// * you need to add an entry to the `render` enum to render the error message as a template
-/// * you need to add an entry to the `py_dict` enum to generate `ctx` for error messages
-#[derive(Clone, Debug, Display, EnumMessage, EnumIter)]
-#[strum(serialize_all = "snake_case")]
-pub enum ErrorType {
+fn do_nothing<T>(v: T) -> T {
+    v
+}
+
+macro_rules! basic_error_default {
+    (
+        $item:ident $(,)?
+    ) => {
+        pub const $item: ErrorType = ErrorType::$item { context: None };
+    };
+    (
+        $item:ident, $($key:ident),* $(,)?
+    ) => {}; // With more parameters enum item must be explicitly created
+}
+
+macro_rules! error_types {
+    (
+        $(
+            $item:ident {
+                $($key:ident: {ctx_type: $ctx_type:ty, ctx_fn: $ctx_fn:path, extract_type: $extract_type:ty}),* $(,)?
+            },
+        )+
+    ) => {
+        #[derive(Clone, Debug, Display, EnumMessage, EnumIter)]
+        #[strum(serialize_all = "snake_case")]
+        pub enum ErrorType {
+            $(
+                $item {
+                    context: Option<Py<PyDict>>,
+                    $($key: $ctx_type,)*
+                }
+            ),+,
+        }
+        impl ErrorType {
+            pub fn new(py: Python, value: &str, context: Option<&PyDict>) -> PyResult<Self> {
```

Comment:
I think this could be simplified a bit (from original) so the code size would be smaller. Ie the context extraction would be moved to its own `fn`s

Reply:
Did it